pkg/registry/comparison/comparison.go

```diff
@@ -0,0 +1,182 @@
+package comparison
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/caicloud/nirvana/log"
+	"github.com/kleveross/klever-model-registry/pkg/common"
+	"github.com/kleveross/klever-model-registry/pkg/registry/errors"
+	"github.com/kleveross/klever-model-registry/pkg/registry/harbor"
+	"github.com/kleveross/klever-model-registry/pkg/registry/paging"
+	"github.com/kleveross/klever-model-registry/pkg/util"
+	ormbmodel "github.com/kleveross/ormb/pkg/model"
+)
+
+// ComparisonGenerator generator model list and
+// compare model diff
+func ComparisonGenerator(ctx context.Context, models Comparison, opt *paging.ListOption) (*paging.Result, error) {
+	metaList, err := composeComparison(models.Model)
+	if err != nil {
+		return nil, err
+	}
+	return paging.Page(metaList, opt), nil
+}
+
+func composeComparison(models []ComparisonModel) ([]ormbmodel.Model, error) {
+	proxy := harbor.NewProxy(common.ORMBDomain, common.ORMBUserName, common.ORMBPassword)
+	metaList := make([]ormbmodel.Model, 0)
+	for _, model := range models {
+		artifacts, err := proxy.ListArtifacts(model.Project, model.Name)
+		if err != nil {
+			return nil, err
+		}
+		for _, artifact := range artifacts {
+			for _, tag := range artifact.Tags {
+				if tag.Name == model.Tag {
+					manifest, err := json.Marshal(artifact.ExtraAttrs)
+					if err != nil {
+						return nil, err
+					}
+					var meta ormbmodel.Metadata
+					json.Unmarshal(manifest, &meta)
+					metaList = append(metaList, ormbmodel.Model{
+						Path:     fmt.Sprintf("%s/%s:%s", model.Project, model.Name, model.Tag),
+						Metadata: &meta,
+					})
+					break
+				}
+			}
+		}
+	}
+
+	return metaList, nil
+}
+
+func DownloadCSVFile(ctx context.Context, models Comparison) error {
+	metaList, err := composeComparison(models.Model)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	err = createCSVFile(ctx, metaList)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+func createCSVFile(ctx context.Context, metas []ormbmodel.Model) error {
+	fileName := fmt.Sprintf("%d.csv", time.Now().Unix())
+
+	file, err := os.Create(fileName)
+	if err != nil {
+		return fmt.Errorf("open file is failed, err: %s", err.Error())
+	}
+
+	defer func() {
+		file.Close()
+		os.Remove(fileName)
+	}()
+
+	// for UTF-8 BOM
+	_, err = file.WriteString("\xEF\xBB\xBF")
+	if err != nil {
+		return fmt.Errorf("write utf-8 bom err: %s", err.Error())
+	}
+	csvWrite := csv.NewWriter(file)
+
+	content, err := composeCSVFileContent(metas)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	err = csvWrite.WriteAll(content)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	csvWrite.Flush()
+
+	file, err = os.Open(fileName)
+	if err != nil {
+		return fmt.Errorf("open file is failed, err: %s", err.Error())
+	}
+
```

Comment:
Need `defer file.Close()` or not?

Reply:
See line 84

Reply:
So why `os.Open(fileName) ` again,  want to seek to begin?

Reply:
There need twice  Close()? First is be for os.Create, the second is for os.Open

Reply:
Only need to close once. See the error message below when close twice:

```
ERROR 0915-11:22:01.907+08 comparison.go:88 | close file is failed, err: %sclose 1600140121.csv: file already closed
```

Reply:
Use file.Seek instead of open file

Reply:
lgtm